lib/grant-types/authorization-code-grant-type.js

```diff
@@ -0,0 +1,170 @@
+    throw new InvalidArgumentError('Invalid argument: model does not implement `revokeAuthorizationCode()`');
+  }
+
+  if (!options.model.saveToken) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveToken()`');
+  }
+
+  AbstractGrantType.call(this, options);
+}
+
+/**
+ * Inherit prototype.
+ */
+
+util.inherits(AuthorizationCodeGrantType, AbstractGrantType);
+
+/**
+ * Handle authorization code grant.
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+ */
+
+AuthorizationCodeGrantType.prototype.handle = function(request, client) {
+  if (!request) {
+    throw new InvalidArgumentError('Missing parameter: `request`');
+  }
+
+  if (!client) {
+    throw new InvalidArgumentError('Missing parameter: `client`');
+  }
+
+  var scope = this.getScope(request);
+
+  return Promise.bind(this)
+    .then(function() {
+      return this.getAuthorizationCode(request, client);
+    })
+    .tap(function(code) {
+      return this.revokeAuthorizationCode(code);
+    })
+    .then(function(code) {
+      return this.saveToken(code.user, client, code.authorizationCode, scope);
+    });
+};
+
+/**
+ * Get the authorization code.
+ */
+
+AuthorizationCodeGrantType.prototype.getAuthorizationCode = function(request, client) {
+  if (!request.body.code) {
+    throw new InvalidRequestError('Missing parameter: `code`');
+  }
+
+  if (!is.vschar(request.body.code)) {
+    throw new InvalidRequestError('Invalid parameter: `code`');
+  }
+
+  return Promise.try(this.model.getAuthorizationCode, request.body.code)
+    .then(function(code) {
+      if (!code) {
+        throw new InvalidGrantError('Invalid grant: authorization code is invalid');
+      }
+
+      if (!code.client) {
+        throw new ServerError('Server error: `getAuthorizationCode()` did not return a `client` object');
+      }
+
+      if (!code.user) {
+        throw new ServerError('Server error: `getAuthorizationCode()` did not return a `user` object');
+      }
+
+      if (code.client.id !== client.id) {
+        throw new InvalidGrantError('Invalid grant: authorization code is invalid');
+      }
+
+      if (!(code.expiresAt instanceof Date)) {
+        throw new ServerError('Server error: `expiresAt` must be a Date instance');
+      }
+
+      if (code.expiresAt < new Date()) {
+        throw new InvalidGrantError('Invalid grant: authorization code has expired');
+      }
+
+      return code;
+    });
+};
+
+/**
+ * Revoke the authorization code.
+ *
+ * "The authorization code MUST expire shortly after it is issued to mitigate
+ * the risk of leaks. [...] If an authorization code is used more than once,
+ * the authorization server MUST deny the request."
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-4.1.2
+ */
+
+AuthorizationCodeGrantType.prototype.revokeAuthorizationCode = function(code) {
+  return Promise.try(this.model.revokeAuthorizationCode, code)
+    .then(function(code) {
+      if (!code) {
```

Comment:
In fact, it doesn't. I can remove it if you do not find that it is helpful.

I added this to force the user's callback to return an expired code, as you can see from the check a [few lines below](https://github.com/thomseddon/node-oauth2-server/pull/203/files#diff-99565ee28a520ab376c88b2fdefffeb6R140).

This was meant from the point of view of "enforcing" the OAuth spec by making sure that the code was revoked and that the user's implementation of the function did not fail. (of course, I understand that there are a myriad of ways to cheat this, as with anything)